### PR TITLE
Add `meshgrid` subroutine in `stdlib_math`

### PR DESCRIPTION
**Pull request by Ivanou34**
WIP: first_proposal

The PR implements `meshgrid`, based on the API described in #18. It includes tests, spec and examples.

All versions for `integer` and `real` types are generated with fypp, as well as multi-dimensional versions, from 1 up to the maximum rank allowed.

The use of the subroutine `error_stop` in the case of invalid argument `indexing` prevented me to declare the subroutine with the `pure` attribute.

### CODE REVIEW

File: doc/specs/stdlib_math.md [hunk 1]
```diff
@@ -554,3 +554,59 @@ When both `prepend` and `append` are not present, the result `y` has one fewer e
...
+For a 2D problem in Cartesian indexing:
+`call [[stdlib_math(module):meshgrid(interface)]](x, y, xm, ym)`
+
+For a 3D problem in Cartesian indexing:
+`call [[stdlib_math(module):meshgrid(interface)]](x, y, z, xm, ym, zm)`
+
+For a 3D problem in matrix indexing:
+`call [[stdlib_math(module):meshgrid(interface)]](x, y, z, xm, ym, zm, indexing="ij")`
+
+The subroutine can be called in $n$-dimensional situations, as long as $n$ is inferior to the maximum allowed array rank.
```
Comment (jvdp1) [inline doc/specs/stdlib_math.md:579]:
```suggestion
The subroutine can be called in `n`-dimensional situations, as long as `n` is inferior to the maximum allowed array rank.
```

File: doc/specs/stdlib_math.md [hunk 2]
```diff
@@ -554,3 +554,59 @@ When both `prepend` and `append` are not present, the result `y` has one fewer e
...
+#### Arguments
+
+For a `n`-dimensional problem, with `n >= 1`:
+
+`x1, x2, ..., xn`: The coordinate vectors.
+Shall be `real/integer` and `rank-1` arrays.
+These arguments are `intent(in)`.
+
+`xm1, xm2, ..., xmn`: The coordinate matrices.
+Shall be `real/integer` arrays of adequate shape:
```
Comment (jvdp1) [inline doc/specs/stdlib_math.md:598]:
```suggestion
Shall be arrays of type `real` or `integer` of adequate shape:
```

File: doc/specs/stdlib_math.md [hunk 3]
```diff
@@ -554,3 +554,59 @@ When both `prepend` and `append` are not present, the result `y` has one fewer e
...
+- for Cartesian indexing, the shape of the coordinate matrices must be `[size(x2), size(x1), size(x3), ..., size(xn)]`.
+- for matrix indexing, the shape of the coordinate matrices must be `[size(x1), size(x2), size(x3), ..., size(xn)]`.
+
+These argument are `intent(out)`.
+
+`indexing`: the selected indexing.
+Shall be a `character(len=2)` equal to `"xy"` for Cartesian indexing (default), or `"ij"` for matrix indexing.
```
Comment (jvdp1) [inline doc/specs/stdlib_math.md:605]: Suggestion: provide an `integer` `paramete`, e.g.
```fortran
integer, parameter, public :: stdlib_meshgrid_xy = 0 , stdlib_meshgrid_ij = 1
```

File: src/stdlib_math_meshgrid.fypp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      #:set XY_INDICES = [1]
+      #:set IJ_INDICES = [1]
+    #:else
+      #:set XY_INDICES = [2, 1] + [j for j in range(3, rank + 1)]
+      #:set IJ_INDICES = [1, 2] + [j for j in range(3, rank + 1)]
+    #:endif
+    #: set RName = rname("meshgrid", rank, t1, k1)
+    module procedure ${RName}$
+        use stdlib_optval, only: optval
+        use stdlib_error, only: error_stop
```
Comment (jvdp1) [inline src/stdlib_math_meshgrid.fypp:34]: Should be moved to the top, just after the line 18.

File: test/math/test_meshgrid.fypp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    subroutine test_${RName}$(error)
+        !> Error handling
+        type(error_type), allocatable, intent(out) :: error
+        integer, parameter :: length = 3
+        ${t1}$ :: ${"".join(f"x{j}(length)," for j in range(1, rank + 1)).removesuffix(",")}$
+        ${t1}$ :: ${"".join(f"xm{j}({GRIDSHAPE})," for j in range(1, rank + 1)).removesuffix(",")}$
+        ${t1}$ :: ${"".join(f"xm{j}_exact({GRIDSHAPE})," for j in range(1, rank + 1)).removesuffix(",")}$
+        integer :: i
+        integer :: ${"".join(f"i{j}," for j in range(1, rank + 1)).removesuffix(",")}$
+        ${t1}$, parameter :: ZERO = 0
```
Comment (jvdp1) [inline test/math/test_meshgrid.fypp:66]: is `0` not too strict? is it not prone to false failures?

**Review by jvdp1 — COMMENTED** [review]
Thank you for this PR.
I am not familiar with this. However, the specs, code and tests seem to be good to me. I have only a few suggestions.

File: test/math/test_meshgrid.fypp (hunk 5 above)
Comment (Ivanou34) [reply test/math/test_meshgrid.fypp:66]: For `integer` types the no. For `real` types the `testdrive` subroutine `check` has a default threshold for equality comparison equal to `epsilon(expected)` (with `expected=ZERO` our case). All tests succeed on my laptop, but if recommended I can specify a bigger threshold.

File: doc/specs/stdlib_math.md (hunk 3 above)
Comment (Ivanou34) [reply doc/specs/stdlib_math.md:605]: Ok I like it

File: src/stdlib_math_meshgrid.fypp (hunk 4 above)
Comment (Ivanou34) [reply src/stdlib_math_meshgrid.fypp:34]: Yes indeed

File: doc/specs/stdlib_math.md [hunk 6]
```diff
@@ -554,3 +554,59 @@ When both `prepend` and `append` are not present, the result `y` has one fewer e
...
+
+Computes a list of coordinate matrices from coordinate vectors.
+
+For $n \geq 1$ coordinate vectors $(x_1, x_2, ..., x_n)$ of sizes $(s_1, s_2, ..., s_n)$, `meshgrid` computes $N$ coordinate matrices $(X_1, X_2, ..., X_n)$ with identical shape corresponding to the selected indexing:
+- Cartesian indexing (default behavior): the shape of the coordinate matrices is $(s_2, s_1, s_3, s_4, ... s_n)$.
+- matrix indexing: the shape of the coordinate matrices is $(s_1, s_2, s_3, s_4, ... s_n)$.
+
+#### Syntax
+
+For a 2D problem in Cartesian indexing:
```
[the remark this answers is not in the document]
Comment (Ivanou34) [reply doc/specs/stdlib_math.md:570]: Mmmh I am not sure to understand the question. There is a difference between matrix and Cartesian indexing in 2D (the shape of the result is not even the same). In 1D there is no difference, but in this case there is no problem at all.

File: test/math/test_meshgrid.fypp (hunk 5 above)
Comment (jvdp1) [reply test/math/test_meshgrid.fypp:66]: Thank you. LGTM!

**Review by jvdp1 — APPROVED** [review]
LGTM. Thank you @Ivanou34 for this PR.

**Review by perazz — APPROVED** [review]
LGTM, Thank you @Ivanou34

**Comment by jvdp1** [comment]
Thank you @perazz for your review. I will fix the conflict ( in a CMake file) and merge it.
Thank you @Ivanou34 for this PR!

**Pull request merged**